crates/fast_compute/src/memory.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright 2025 Irreducible Inc.
+
+use std::marker::PhantomData;
+
+use binius_compute::memory::{ComputeMemory, SizedSlice};
+use binius_field::PackedField;
+
+/// A packed memory implementation that uses slices of packed fields.
+pub struct PackedMemory<P>(PhantomData<P>);
+
+impl<P: PackedField> ComputeMemory<P::Scalar> for PackedMemory<P> {
+	const MIN_SLICE_LEN: usize = P::WIDTH;
+
+	type FSlice<'a> = PackedMemorySlice<'a, P>;
+
+	type FSliceMut<'a> = PackedMemorySliceMut<'a, P>;
+
+	fn as_const<'a>(data: &'a Self::FSliceMut<'_>) -> Self::FSlice<'a> {
+		PackedMemorySlice { data: data.data }
+	}
+
+	fn slice(data: Self::FSlice<'_>, range: impl std::ops::RangeBounds<usize>) -> Self::FSlice<'_> {
+		let (start, end) = Self::to_packed_range(data.len(), range);
+		Self::FSlice {
+			data: &data.data[start..end],
+		}
+	}
+
+	fn slice_mut<'a>(
+		data: &'a mut Self::FSliceMut<'_>,
+		range: impl std::ops::RangeBounds<usize>,
+	) -> Self::FSliceMut<'a> {
+		let (start, end) = Self::to_packed_range(data.len(), range);
+
+		Self::FSliceMut {
+			data: &mut data.data[start..end],
+		}
+	}
+
+	fn split_at_mut(
+		data: Self::FSliceMut<'_>,
+		mid: usize,
+	) -> (Self::FSliceMut<'_>, Self::FSliceMut<'_>) {
+		assert_eq!(mid % P::WIDTH, 0, "mid must be a multiple of {}", P::WIDTH);
+		let mid = mid >> P::LOG_WIDTH;
+		let (left, right) = data.data.split_at_mut(mid);
+		(Self::FSliceMut { data: left }, Self::FSliceMut { data: right })
+	}
+
+	fn narrow<'a>(data: &'a Self::FSlice<'_>) -> Self::FSlice<'a> {
+		Self::FSlice { data: data.data }
+	}
+
+	fn slice_chunks_mut<'a>(
+		data: Self::FSliceMut<'a>,
+		chunk_len: usize,
+	) -> impl Iterator<Item = Self::FSliceMut<'a>> {
+		assert_eq!(chunk_len % P::WIDTH, 0, "chunk_len must be a multiple of {}", P::WIDTH);
+		assert_eq!(data.len() % chunk_len, 0, "data.len() must be a multiple of chunk_len");
+
+		let chunk_len = chunk_len >> P::LOG_WIDTH;
+
+		data.data
+			.chunks_mut(chunk_len)
+			.map(|chunk| Self::FSliceMut { data: chunk })
+	}
+}
+
+impl<P: PackedField> PackedMemory<P> {
+	fn to_packed_range(len: usize, range: impl std::ops::RangeBounds<usize>) -> (usize, usize) {
+		let start = match range.start_bound() {
+			std::ops::Bound::Included(&start) => start,
+			std::ops::Bound::Excluded(&start) => start + 1,
```

Comment:
Should 1 be P::WIDTH?

Reply:
TBH I doubt how that should work correctly but changed to P::WIDTH